rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -78,20 +88,25 @@ class rmw_context_impl_s::Data final
     }
 
     // Check if shm is enabled.
-    z_owned_str_t shm_enabled = zc_config_get(z_loan(config), "transport/shared_memory/enabled");
+    z_owned_string_t shm_enabled;
+    zc_config_get_from_str(z_loan(config), Z_CONFIG_SHARED_MEMORY_KEY, &shm_enabled);
     auto always_free_shm_enabled = rcpputils::make_scope_exit(
       [&shm_enabled]() {
         z_drop(z_move(shm_enabled));
       });
 
     // Initialize the zenoh session.
-    session_ = z_open(z_move(config));
-    if (!z_session_check(&session_)) {
+    if (z_open(&session_, z_move(config), NULL) != Z_OK) {
+      RMW_SET_ERROR_MSG("Error setting up zenoh session");
       throw std::runtime_error("Error setting up zenoh session.");
     }
+    atexit(update_is_exiting);
     auto close_session = rcpputils::make_scope_exit(
       [this]() {
-        z_close(z_move(session_));
+        // Don't touch Zenoh Session if the ROS process is exiting, it will cause panic.
```

Comment:
In what scenarios did you observe a crash? Is it still a problem if users shutdown the ROS context as we explain here https://github.com/ros2/rmw_zenoh#crash-when-program-terminates?

Reply:
Let's also discuss it here https://github.com/ros2/rmw_zenoh/issues/324